pytorch_lightning/utilities/fetching.py

```diff
@@ -174,35 +208,183 @@ class DataFetcher(AbstractDataFetcher):
 
     """
     This class is used to control batch fetching flow.
+
+    By default, the `fetching_function` will `prefetch` a batch in advance to detect the end of the iteration.
     """
 
+    @contextlib.contextmanager
+    def fetching_context(self):
+        """Hook to override to add context logic around batch fetching"""
+        yield
+
+    def on_fetch_start(self) -> None:
+        """Hook to override to handle the logic before fetching a batch"""
```

Comment:
i'm confused - should users be extending `DataFetcher` or implementing `AbstractDataFetcher` if this hook is meant to be overridden? should this be enforced at the interface level?

Reply:
The `AbstractDataFetcher` should be `subclassed` for users to build their own CustomDataFetcher. 

However, as implementing properly a fetcher can be cumbersome,
the `DataFetcher` also provide some hooks to customise its behaviour.
Right now, those hooks are tailored for the `InterBatchParallelismDataFetcher` which is using those hooks to wrap fetching under Cuda Context and `on_fetch_start` and `on_fetch_end` to handle event recording logic.
I believe this could be useful if this logic can be extended for TPU too.

Reply:
> However, as implementing properly a fetcher can be cumbersome, the DataFetcher also provide some hooks to customise its behaviour. Right now, those hooks are tailored for the InterBatchParallelismDataFetcher which is using those hooks to wrap fetching under Cuda Context and on_fetch_start and on_fetch_end to handle event recording logic. I believe this could be useful if this logic can be extended for TPU too.

IMO, we should stick to using the AbstractDataFetcher as the pure interface for these. Otherwise we're in this mix of inheritance (extending DataFetcher) vs implementation (implementing AbstractDataFetcher). The codebase can quickly become inconsistent as some places assume DataFetcher vs other places assuming AbstractDataFetcher. 

and people get confused why there are multiple base classes for them to work with :/